### PR TITLE
GH-475: Optional Feign clients eager initialization mechanism

### LINKED ISSUE #475: Feign client lazy initialization fails when using CompletableFuture.supplyAsync() to call the Feign client initially
**Describe the bug**
Feign client initialization is always lazy. As such, the first calling thread triggers the initialization, which may be another Spring component (e.g. REST controller) using `CompletableFuture.supplyAsync()` to do so. This uses `ForkJoinWorkerThread` underneath, resulting in a `ClassNotFoundException`, if and only if running in a dockerized environment based on Spring Boot Build Image plugin.

Using other JVM Executors(except `newWorkStealingExecutor)`, e.g. `fixedSizeThreadExecutor()`, does not result in this behavior. Also eagerly initializing the Feign client by calling it synchronously, e.g. from within a `CommandLineRunner`, circumvents this. Once, the Feign client is properly initialized, there is no problem, however, Feign does not yet support eager initialization.

**Sample**
A sample project with instructions to reproduce is provided here: https://github.com/maverick1601/openfeign-build-image-error

### PR DESCRIPTION
**Pull request by Periecle**
Pull Request majorly influenced by[ similar Ribbon feature](https://github.com/spring-cloud/spring-cloud-netflix/pull/1749), therefore API and configuration are close to each other for usability. 
Also updated documentation explaining configuration and implementation details. 
Fixes #475

### CODE REVIEW

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignAutoConfiguration.java [hunk 1]
```diff
@@ -301,4 +302,18 @@ public RequestInterceptor oauth2FeignRequestInterceptor(OAuth2ClientContext oAut
...
+	@Configuration(proxyBeanMethods = false)
+	@ConditionalOnProperty(value = "feign.eager-load.enabled", havingValue = "true")
+	protected static class FeignEagerLoadConfiguration {
+
+		@Autowired(required = false)
+		private FeignEagerLoadProperties feignEagerLoadProperties;
+
+		@Bean
+		public FeignApplicationContextInitializer feignApplicationContextInitializer(FeignContext feignContext) {
+			return new FeignApplicationContextInitializer(feignContext, feignEagerLoadProperties.getClients());
```
[the remark this answers is not in the document]
Comment (OlgaMaciaszek) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignAutoConfiguration.java:314]: On a second thought, it does not seem there's a need to autowire it this way and mark as `required=false`. We could just inject it as an argument instead:
```
@Bean
		public FeignApplicationContextInitializer feignApplicationContextInitializer(FeignContext feignContext,
				FeignEagerLoadProperties feignEagerLoadProperties) {
			return new FeignApplicationContextInitializer(feignContext, feignEagerLoadProperties.getClients());
		}
```

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
@Periecle thanks for submitting this. Have added some comments to address, however, have also run some tests on the sample provided in the issue (https://github.com/spring-cloud/spring-cloud-openfeign/issues/475) and this does not really seem to solve the issue, so probably will require further investigation.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignAutoConfiguration.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Periecle) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignAutoConfiguration.java:314]: Agree, properties can be injected this way, will do it.

File: spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignEagerLoadProperties.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ * should be eagerly loaded up
+ *
+ * @author Roman Kvasnytskyi
+ */
+@ConfigurationProperties(prefix = "feign.eager-load")
+public class FeignEagerLoadProperties {
+
+	private boolean enabled = false;
+
+	private List<String> clients;
```
[the remark this answers is not in the document]
Comment (Periecle) [reply spring-cloud-openfeign-core/src/main/java/org/springframework/cloud/openfeign/FeignEagerLoadProperties.java:34]: Okay, will do

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientEagerInitializationTests.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	protected interface LazyFeignClient {
+
+		@RequestMapping(method = RequestMethod.GET, value = "/hello")
+		Hello getHello();
+
+	}
+
+	@Configuration(proxyBeanMethods = false)
+	@EnableAutoConfiguration
+	@EnableConfigurationProperties(FeignEagerLoadProperties.class)
```
[the remark this answers is not in the document]
Comment (Periecle) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignClientEagerInitializationTests.java:96]: Reasonable. Will fix this

**Comment by Periecle** [comment]
> @Periecle thanks for submitting this. Have added some comments to address, however, have also run some tests on the sample provided in the issue (#475) and this does not really seem to solve the issue, so probably will require further investigation.

Author of original issue was struggling with initialising  feign clients in async way from ForkJoinPool, you proposed to implement something similar to ribbon.eager-load mechanism. I think my PR won't fix sample from original issue, but propose options to provide such functionality directly without need for workarounds. 
On second thought, it seems that cloud-openfeign need option for initialising Ribbon LB clients with something like 
feign.ribbon.warmup. Actually, I am already working on this feature.

**Comment by Periecle** [comment]
@OlgaMaciaszek, hi. Can you update me with current status of review?

**Comment by OlgaMaciaszek** [comment]
> hi. Can you update me with current status of review?

@Periecle , I understood you were still working on this bit for the PR:

> On second thought, it seems that cloud-openfeign need option for initialising Ribbon LB on ApplicationReadyEvent optionally on feign.ribbon.warmup, . Actually, I am already working on this feature.

Or will you be submitting it in another PR?

**Pull request closed without merging**